examples/h_mountain.py

```diff
@@ -0,0 +1,179 @@
+from gusto import *
+from firedrake import FunctionSpace, as_vector, \
+    VectorFunctionSpace, PeriodicIntervalMesh, ExtrudedMesh, \
+    SpatialCoordinate, exp, pi, cos, Function, conditional, Mesh, sin, op2, sqrt
+import sys
+
+dt = 5.0
+if '--running-tests' in sys.argv:
+    tmax = dt
+else:
+    tmax = 15000.
+
+res = 10
+nlayers = res*20 # horizontal layers
+columns = res*12 # number of columns
+L = 240000.
+m = PeriodicIntervalMesh(columns, L)
+
+# build volume mesh
+H = 50000. # Height position of the model top
+ext_mesh = ExtrudedMesh(m, layers=nlayers, layer_height=H/nlayers)
+Vc = VectorFunctionSpace(ext_mesh, "DG", 2)
+coord = SpatialCoordinate(ext_mesh)
+x = Function(Vc).interpolate(as_vector([coord[0], coord[1]]))
+a = 10000.
+xc = L/2.
+x, z = SpatialCoordinate(ext_mesh)
+hm = 1.
+zs = hm*a**2/((x-xc)**2 + a**2)
+smooth_z = True
+if smooth_z:
+    zh = 5000.
+    xexpr = as_vector([x, conditional(z < zh, z + cos(0.5*pi*z/zh)**6*zs, z)])
+else:
+    xexpr = as_vector([x, z + ((H-z)/H)*zs])
+new_coords = Function(Vc).interpolate(xexpr)
+mesh = Mesh(new_coords)
+
+# sponge function
+W_DG = FunctionSpace(mesh, "DG", 2)
+x, z = SpatialCoordinate(mesh)
+zc = H-20000.
+mubar = 0.3/dt
+mu_top = conditional(z <= zc, 0.0, mubar*sin((pi/2.)*(z-zc)/(H-zc))**2)
+mu = Function(W_DG).interpolate(mu_top)
+fieldlist = ['u', 'rho', 'theta']
+timestepping = TimesteppingParameters(dt=dt, alpha=1.0)
+output = OutputParameters(dirname='h_mountain_smootherz', dumpfreq=30, dumplist=['u'], perturbation_fields=['theta', 'rho'])
+parameters = CompressibleParameters(g=9.80665, cp=1004.)
+diagnostics = Diagnostics(*fieldlist)
+diagnostic_fields = [CourantNumber(), VelocityZ(), HydrostaticImbalance()]
+
+state = State(mesh, vertical_degree=1, horizontal_degree=1,
+              family="CG",
+              sponge_function=mu,
+              hydrostatic=True,
+              timestepping=timestepping,
+              output=output,
+              parameters=parameters,
+              diagnostics=diagnostics,
+              fieldlist=fieldlist,
+              diagnostic_fields=diagnostic_fields)
+
+
+# Off-centring parameter is set to alpha=1.0 for first time step
+# to omit spurious orography waves, then alpha=0.5
+def implicit_first_step():
+    if state.t.dat.data_ro[0] == dt:
```

Comment:
Careful comparing floating point values exactly.  You might be OK here, but it could be incidental.

Reply:
I think this should be ok. After the first time step we have ```state.t.dat.data_ro[0] = float(dt)```, and python interprets ```float(dt) == dt``` as true both for dt as an integer and as a float.

Reply:
That's not what I mean.  Consider:

```python
dt = 0.1
x = 0
for i in range(10):
    x += dt
print(x == dt*10) => False
```

Reply:
Hmm interesting, thanks for pointing this out! I wasn't aware of this issue before. In this case here it should still be ok since I think ```0+dt == dt``` is true for any time increment value dt. If the initial time isn't 0 this need not be the case anymore, so maybe in this case one should write something like
```
t_init = 0

def implicit_first_step(time):
    if time >= t_init + dt:
        state.timestepping.alpha = 0.5

stepper.run(t=t_init, tmax=tmax)
```